src/nn/embedding.rs

```diff
@@ -129,6 +135,134 @@ impl<const VOCAB: usize, const DIM: usize, D1: Device<f32>, D2: Device<f32>> ToD
     }
 }
 
+/// A learned positional embedding
+/// Initializes [Self::weight] from a Uniform distribution
+/// between [-1 / sqrt(I), 1 / sqrt(I)].
+///
+/// # Generics
+/// - `MAX_LEN` The maximum length of the input sequences. This determines how many positional embeddings to train.
+/// - `DIM` The "output" size of vectors & matrices which are the vectors being selected.
+///
+/// # Examples
+/// `LearnedPositionalEmbedding<5, 2>` can act on vectors with SEQ x DIM elements (pre-embedded),
+/// resulting in output vectors of the same size with positional embeddings added.
+///
+/// ```rust
+///
+/// # use dfdx::prelude::*;
+/// # let dev: Cpu = Default::default();
+/// type Model = (Embedding<7, 2>, LearnedPositionalEmbedding<20, 2>);
+/// let mut model = Model::build_on_device(&dev);
+/// // single sequence of ids
+/// let inputs: Tensor<Rank1<5>, usize, _> = dev.zeros();
+/// let _: Tensor<(Const<5>, Const<2>,), f32, _> = model.forward(inputs);
+/// // batched sequence of ids
+/// let inputs: Tensor<Rank2<10, 5>, usize, _> = dev.zeros();
+/// let _: Tensor<(Const<10>, Const<5>, Const<2>), f32, _> = model.forward(inputs);
+/// ```
+#[derive(Debug, Clone)]
+pub struct LearnedPositionalEmbedding<
+    const MAX_LEN: usize,
+    const DIM: usize,
+    E: Dtype,
+    D: DeviceStorage,
+> {
+    /// Learned positonal embeddings
+    embed: Embedding<MAX_LEN, DIM, E, D>,
+    device: D,
+}
+
+/// Pass in an unbatched pre-embedded sequence, add positional embeddings in
+impl<const MAX_LEN: usize, const DIM: usize, SEQ: Dim, D: Device<f32>, T: Tape<D>>
+    Module<Tensor<(SEQ, Const<DIM>), f32, D, T>>
+    for LearnedPositionalEmbedding<MAX_LEN, DIM, f32, D>
+where
+    D: TensorFrom<std::vec::Vec<usize>, (SEQ,), usize>,
+{
+    type Output = Tensor<(SEQ, Const<DIM>), f32, D, T>;
+    fn forward(&self, input: Tensor<(SEQ, Const<DIM>), f32, D, T>) -> Self::Output {
+        let (input, tape) = input.split_tape();
+        let positions: Tensor<(SEQ,), usize, D> = self
+            .device
+            .tensor((0..input.shape().0.size()).collect::<std::vec::Vec<_>>());
```

Comment:
Nice 😀 Do you need the vec generic parameter to collect? Would've thought rust could infer that